### ISSUE TITLE
Some questions on hyperparameter optimization

### ISSUE THREAD

**Issue by saifsiddique** [chiahao3/ptyrad#19]
For the hyperparameter optimization:

1. you say we don't necessarily need to optimize for `conv_angle` and `defocus`. How close to the actual values do I need to be (for defocus) in the `.yml` files for the reconstruction code to still work?
2. sample thickness optimization is controlled by `dz` and `Nlayer` under `hypertune_params`. Your example `.yml` files optimize for `dz`. In addition to the error values generated, how to determine (by looking at the tiff stacks) if a particular value of `dz` is the right one?
3. some of the trials during the optimization are failing or getting pruned. What might be the reason for these trails to fail/prune and how to minimize this?

**Comment by chiahao3** [comment]
Hi @saifsiddique ,

Thanks for trying out PtyRAD!

1. The reason behind "You don't necessarily need to optimize convergence angle and defocus with hypertune" is because the AD would automatically optimize the probe as a whole, so it's arguably more efficient than spinning multiple trials to test different convergence angles / defocus. In addition, we should "calibrate" convergence angle, instead of trying to change it for each ptycho dataset as mentioned in issue #17. For high quality dataset (i.e., high dose, reasonable k-space sampling), thin samples like the tBL-WSe2 can converge easily with a +- 20nm defocus by just AD itself, but I think for normal specimen thickness you'll want the defocus to be +-5nm. For low dose dataset like the MOSS-6 and ZSM-5 (1e3 e-/Ang^2 or lower), the probe got barely updated with AD so you'll want your initial defocus to be "as close as possible". That being said, there're still benefit of using hypertune to experiment some initial conditions, although I'd often argue it's probably more efficient to try to either calibrate or estimate these values (like aberrations) better before running hypertune.

2. Sample thickness can be optimized by AD (via variable dz) and/or hypertune (changing the initial value of Nlayer or dz). Error values aren't particularly informative unless it's obviously off. I'd recommend keeping your dz to be between 5 - 20 Ang for normal materials. 10 Ang is usually very good unless you're working with very heavy elements, or very thin sample like 2D materials then you'll want thinner slices. The choice of dz is mostly determined by kV and atomic number, and the key is to prevent phase wrapping within a slice, so if your object phase within the slice is approaching 1 radian, you'll likely need to reduce the slice thickness. Often times if the Nlayer would need to be tested with trial-and-error and we usually estimate it with PACBED or EELS and then slowly increase the ptycho object total thickness and see how it goes.

3. The pruner will prune a specific trial because the returned iteration error (i.e., the loss) isn't very promising. So basically if the pruner think this trial is not looking great it'll prune it early to save the resource. This an optional feature, and you're free to turn it off by setting `'pruner_params'  : null`. To fine tune the pruner behavior, I'll recommend you check out the Optuna doc [here](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.HyperbandPruner.html#optuna.pruners.HyperbandPruner). There're other type of pruners as well, presumably you may want to check out ThresholdPruner or PercentilePruner. Just note that pruning, or the entire hypertune workflow would only make sense if the error metric is sensitive to your optimization task. So if I try to use hypertune to finetune defocus on the tBL-WSe2 dataset, it'll likely get very confused and useless because +-20nm of defocus all seem to give similar error after certain amount of iterations.

**Comment by chiahao3** [comment]
@saifsiddique There's a small bug about disabling pruner in hypertune mode as mentioned in issue #20 , you'll need to skip the params validation.
